DDP should not set grad for globally unused params

YinglinSun opened the pull request:
#28294 DDP should not set grad for globally unused parameters

DDP currently computes the param to bucket mapping upfront, and allreduce grads for all params in every iteration. Even if params are unused, it will just set grad to zero. With such behavior, optimizer cannot tell if a param indeed has a zero grad or it is not used in the current iteration. This could trigger convergence problems for optimizers with weight decay and momentum such as SGD. However, DDP cannot simply set grad to None for local unused parameters, as local unused parameters might be used in other processes, and hence we still need to allreduce its grad. Instead DDP should figure out the globally unused parameters and skip touching their grad in the end of backward.

Implementation summary:
* Add locally used parameter map for each model replica.
* Mark the locally unused parameters in the end of forward and then reduce to get the globally unused parameters.
* In the end of backward skip touching grad for those globally unused parameters.
* Add a unit test test_global_local_unused_params_grad

baobablyh commented:
Working on adding unit tests.

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -364,8 +366,10 @@ void Reducer::mark_bucket_ready(size_t bucket_index) {
+      local_used.push_back(replica.local_used);
     }
     bucket.work = process_group_->allreduce(tensors);
+    bucket.work = process_group_->allreduce(local_used);
```
Ah, yes, it wouldn't make sense if we coalesce int and float tensors together.

mrshenli commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -115,6 +115,11 @@ class Reducer {
+    // Per-bucket tensor for indicating if variable is used locally during the
+    // current iteration. This tensor will be allreduced for figuring out the
+    // globally unused parameters. 1 means locally used and 0 locally unused.
+    at::Tensor local_used;
```
As it seems it's not easy to coalesce the bitmap into the existing bucket allreduce, I wonder whether we should make this per-backward bitmap as @pietern suggested, so that we don't have to introduce multiple additional allreduce.

This allreduce can even be done at the end of the forward pass in `prepare_for_backward` where we find all unused params. In that way we will have minimum perf impact.

baobablyh replied:
OK let's keep using integer tensor then. int32 should be good enough? as the number of ranks cannot be that big.

mrshenli reviewed (commented):
A general comment is that we might want to add an option (in `DistributedDataParallel` in `parallel/distributed.py`) to enable/disable this mode. And by default it should off. But there is also another concern that we already have too many options in DDP. @pietern any comments?

I would still vote for adding an option for this mode, as it might have perf impact that most use cases don't have to have.

YinglinSun reviewed (commented):
> A general comment is that we might want to add an option (in `DistributedDataParallel` in `parallel/distributed.py`) to enable/disable this mode. And by default it should off. But there is also another concern that we already have too many options in DDP. @pietern any comments?
> 
> I would still vote for adding an option for this mode, as it might have perf impact that most use cases don't have to have.

I'm OK for the option.

pietern reviewed (commented):
The behavior today is broken, so we shouldn't make it optional.

pietern reviewed (commented):
On second thought, we can (almost) completely hide this by using a little header in every bucket of the same dtype as the bucket. We initialize this to 0 and only set it to 1 (in whatever dtype) if the parameter was used. Then when finalizing, we only need to check if the value for this parameter is non-zero (since we summed them in the allreduce call) before assigning it.

baobablyh commented:
> On second thought, we can (almost) completely hide this by using a little header in every bucket of the same dtype as the bucket. We initialize this to 0 and only set it to 1 (in whatever dtype) if the parameter was used. Then when finalizing, we only need to check if the value for this parameter is non-zero (since we summed them in the allreduce call) before assigning it.

I thought about that. It would work for most of tensor types, however may not work for int8/uint8 (overflow) and bool (undefined behavior?).

mrshenli commented:
> It would work for most of tensor types, however may not work for int8/uint8 (overflow) and bool (undefined behavior?).

Yep, we would also need special handling for sparse tensors if we do that. It seems doing it once in `prepare_for_backward` would be terser to implement and maintain? @pietern 

@baobablyh Please add a PR description, it will appear in the git log after land. 

> The behavior today is broken, so we shouldn't make it optional.

This sounds good to me. Let's keep the `DistributedDataParallel` API intact, and only make changes in C++ side then.

baobablyh commented:
> @baobablyh Please add a PR description, it will appear in the git log after land.
Done.
 
> > The behavior today is broken, so we shouldn't make it optional.
> This sounds good to me. Let's keep the `DistributedDataParallel` API intact, and only make changes in C++ side then.
OK.

Commit:
28294: DDP should not set grad for globally unused params

DDP currently computes the param to bucket mapping upfront, and allreduce grads for all params in every iteration. Even if params are unused, it will just set grad to zero. With such behavior, optimizer cannot tell if a param indeed has a zero grad or it is not used in the current iteration. This could trigger convergence problems for optimizers with weight decay and momentum such as SGD. However, DDP cannot simply set grad to None for local unused parameters, as local unused parameters might be used in other processes, and hence we still need to allreduce its grad. Instead DDP should figure out the globally unused parameters and skip touching their grad in the end of backward.

Implementation summary:
* Add locally used parameter map for each model replica.
* Mark the locally unused parameters in the end of forward and then reduce to get the globally unused parameters.
* In the end of backward skip touching grad for those globally unused parameters.
* Add a unit test test_global_local_unused_params_grad

baobablyh commented:
@mrshenli @pietern This PR is ready for review. Could you help take a look? Thanks!

pietern reviewed (commented):
Thanks for your work on this PR, @baobablyh!

We must also consider the `no_sync()` decorator. This can be used to skip gradient synchronization for a few iterations while gradients are accumulated. If we merge this PR as is, it would only set gradients that were touched in the most recent iteration, not older ones. The result is that these gradients retain their local value instead of being set to their globally reduced value.

We can address this by resetting `local_used_maps_` in the `finalize_backward` call. If parameters are used across multiple iterations, then they will correctly toggle their values in this map, and they are assigned as expected when gradient reduction is finally executed.

baobablyh commented:
> Thanks for your work on this PR, @baobablyh!
> 
Thanks a lot for the thorough review. I'll be addressing these comments today.

baobablyh commented on test/test_c10d.py:
```diff
@@ -2433,6 +2433,65 @@ def test_find_unused_parameters(find_unused_parameters, test_default=False):
+        self.assertIsNotNone(t0_p.grad)
+        self.assertIsNotNone(t1_p.grad)
+        self.assertIsNone(task_unused_p.grad)
+
```
What does device mangling mean exactly?

baobablyh replied:
As NCCL does not support allreduce on CPU tensors and gloo supports both CPU and GPU tensors, look like I should use gloo backend for this test case? Just want to make sure I'm doing the right thing.

baobablyh replied:
Done. Now run the test on both CPU and GPU.

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -529,6 +543,10 @@ void Reducer::prepare_for_backward(
     bucket.pending = bucket.replicas.size();
   }
 
+  for (auto& local_used: local_used_maps_) {
```
Will do. Looks like CI clang-tidy does not capture this?

baobablyh replied:
Done.

baobablyh replied:
Is there any plan to add clang-format in CI? Seems clang-tidy does not work well.

pietern replied:
There is #25700 but it turns out it's not as easy. clang-format is different from clang-tidy. The former only deals with text formatting and the latter does static analysis of the code.

baobablyh replied:
Got it. clang-format helps a lot.

Commit:
Capitalize comment for local_used_maps_

baobablyh commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -74,6 +74,12 @@ class Reducer {
 
   bool has_marked_unused_parameters_;
   std::vector<VariableIndex> unused_parameters_;
+  // locally used parameter maps indicating if parameters are used locally
```
Done.

Commit:
Fix format issues detected by clang-format

pietern commented:
@baobablyh What do you think of the other comments? Reasonable?

baobablyh commented:
> @baobablyh What do you think of the other comments? Reasonable?

Yes, addressing them one by one. Will get them done by today.

baobablyh commented:
> Thanks for your work on this PR, @baobablyh!
> 
> We must also consider the `no_sync()` decorator. This can be used to skip gradient synchronization for a few iterations while gradients are accumulated. If we merge this PR as is, it would only set gradients that were touched in the most recent iteration, not older ones. The result is that these gradients retain their local value instead of being set to their globally reduced value.
> 
> We can address this by resetting `local_used_maps_` in the `finalize_backward` call. If parameters are used across multiple iterations, then they will correctly toggle their values in this map, and they are assigned as expected when gradient reduction is finally executed.

Not sure if I'm following. To address such use case, we have to accumulate the locally used params during those no_sync iterations. However, prepare_for_backward is not called in no_sync context. Since the forward output can be different for each iteration in no_sync context, we have to figure out unused parameters for each iteration. That means we have to refactor distributed.py to call prepare_for_backward regardless of no_sync?

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -583,13 +607,23 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
+      const VariableIndex& var_idx = func_[variable.grad_accumulator().get()];
+      bool global_unused =
+        local_used_maps_[var_idx.replica_index][var_idx.variable_index]
+          .item<int>() == 0;
```
Since allreduce of these tensors happens before bucket allreduce and all allreduce are with the same stream, it seems work.wait of the first bucket implies the readiness of these tensors? It may not be good to assume the implicit synchronization though.

baobablyh replied:
Done. Now wait at finalize_backward.

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -570,8 +588,14 @@ void Reducer::prepare_for_backward(
   }
+
+  // allreduce locally used param maps to get the global consensus
+  process_group_->allreduce(local_used_maps_);
```
Agree. Is there a way to register a per iteration (not per param) hook for the moment when we are about to get into backward of this iteration but haven't yet? Per param post-hook works however it is not that convenient to do some per iteration thing such as marking all unused params ready.

baobablyh replied:
Done. Moved it into autograd_hook and reuse existing has_marked_unused_parameters_ to indicating if this allreduce has been kicked off.

Commit:
Address some comments

* Kick off allreduce for local_used_maps_ in autograd_hook
* Wait for allreduce for local_used_maps_ to complete before using it

baobablyh commented:
> > Thanks for your work on this PR, @baobablyh!
> > We must also consider the `no_sync()` decorator. This can be used to skip gradient synchronization for a few iterations while gradients are accumulated. If we merge this PR as is, it would only set gradients that were touched in the most recent iteration, not older ones. The result is that these gradients retain their local value instead of being set to their globally reduced value.
> > We can address this by resetting `local_used_maps_` in the `finalize_backward` call. If parameters are used across multiple iterations, then they will correctly toggle their values in this map, and they are assigned as expected when gradient reduction is finally executed.
> 
> Not sure if I'm following. To address such use case, we have to accumulate the locally used params during those no_sync iterations. However, prepare_for_backward is not called in no_sync context. Since the forward output can be different for each iteration in no_sync context, we have to figure out unused parameters for each iteration. That means we have to refactor distributed.py to call prepare_for_backward regardless of no_sync?

@pietern I think we have to refactor the existing code quite a bit to support this use case. Can we take care of it separately after merging this one? Otherwise the scope of this one may become too involved to be merged soon.

What do you think?

Commit:
Run test_global_local_unused_params_grad on both CPU and GPU

mrshenli commented:
> Not sure if I'm following. To address such use case, we have to accumulate the locally used params during those no_sync iterations. However, prepare_for_backward is not called in no_sync context. Since the forward output can be different for each iteration in no_sync context, we have to figure out unused parameters for each iteration. That means we have to refactor distributed.py to call prepare_for_backward regardless of no_sync?

That's a good question. An alternative might be, instead of calling `prepare_for_backward`, we can have a separate function to mark unused params, which is always called regardless of it's in `no_sync` mode or not. Two reason for doing this might be 1) the name `prepare_for_backward` sounds to suggest backward needs to be called before the next forward; 2) `prepare_for_backward` sets and checks several flags, but what we really need to do in this case is just marking used/unused params. Just a thought.

Commit:
Address the case that some rank does not have unused params

Even if there is no unused params, we should still allreduce local_used_maps_, as other ranks may have unused params waiting for the global consensus.

baobablyh commented:
> > Not sure if I'm following. To address such use case, we have to accumulate the locally used params during those no_sync iterations. However, prepare_for_backward is not called in no_sync context. Since the forward output can be different for each iteration in no_sync context, we have to figure out unused parameters for each iteration. That means we have to refactor distributed.py to call prepare_for_backward regardless of no_sync?
> 
> That's a good question. An alternative might be, instead of calling `prepare_for_backward`, we can have a separate function to mark unused params, which is always called regardless of it's in `no_sync` mode or not. Two reason for doing this might be 1) the name `prepare_for_backward` sounds to suggest backward needs to be called before the next forward; 2) `prepare_for_backward` sets and checks several flags, but what we really need to do in this case is just marking used/unused params. Just a thought.

I like this idea. The logic is much clearer. Would we like to take care of this "no_sync" case separately? I can do it after this PR. Please let me know the thought. I would be OK with either one. Just don't want to block this PR for too long.

mrshenli commented:
> I can do it after this PR. Please let me know the thought. I would be OK with either one. Just don't want to block this PR for too long. Thanks!

+1 for doing it in a separate PR on top of this ([ghstack](https://github.com/ezyang/ghstack)), as long as we are committed to this and make sure we don't break existing use cases. 

@pietern does the plan of having a separate function sounds OK to you?

Commit:
Fix test_forward_backward_unused_parameters

baobablyh commented:
I think I have addressed all comments but 'no_sync'. I also made a few commits for unit test failure and some other issues. I'll squash commits to single one once we are satisfied with the PR.

baobablyh commented:
All failed CI tests seem irrelevant. Some of them were due to this:
`
03:53:49 ../lib/libtorch.so: undefined reference to `void fbgemm::CodeGenBase<unsigned char, signed char, int, int>::initCRegs<(fbgemm::inst_set_t)3>(asmjit::x86::Emitter*, int, int, int)'
03:53:49 ../lib/libtorch.so: undefined reference to `void fbgemm::CodeGenBase<unsigned char, signed char, int, int>::storeCRegs<(fbgemm::inst_set_t)3>(asmjit::x86::Emitter*, int, int, asmjit::x86::Gp, asmjit::x86::Gp, bool, int)'
`

mrshenli commented:
Regarding the `fbgemm` failure, it might relate to the submodule update last might which breaks master. [This](https://github.com/pytorch/pytorch/commit/3b452ca4287ef573b189cefc54192f19da28aef7) is a commit on master passed all tests, maybe rebase to that instead?

baobablyh commented:
> Regarding the `fbgemm` failure, it might relate to the submodule update last might which breaks master. [This](https://github.com/pytorch/pytorch/commit/3b452ca4287ef573b189cefc54192f19da28aef7) is a commit on master passed all tests, maybe rebase to that instead?

Thanks. Trying.

baobablyh commented:
> > Regarding the `fbgemm` failure, it might relate to the submodule update last might which breaks master. [This](https://github.com/pytorch/pytorch/commit/3b452ca4287ef573b189cefc54192f19da28aef7) is a commit on master passed all tests, maybe rebase to that instead?
> 
> Thanks. Trying.

Rebased. Let's see if it helps.

baobablyh commented:
> > > Regarding the `fbgemm` failure, it might relate to the submodule update last might which breaks master. [This](https://github.com/pytorch/pytorch/commit/3b452ca4287ef573b189cefc54192f19da28aef7) is a commit on master passed all tests, maybe rebase to that instead?
> > 
> > 
> > Thanks. Trying.
> 
> Rebased. Let's see if it helps.

Ran into some other issue after rebasing :frowning_face: 

```
The following packages have unmet dependencies:
 linux-image-generic : Depends: linux-image-4.4.0-168-generic but it is not going to be installed or
                                linux-image-unsigned-4.4.0-168-generic but it is not installable
                       Depends: linux-modules-extra-4.4.0-168-generic but it is not installable
                       Recommends: thermald but it is not going to be installed
E: Unable to correct problems, you have held broken packages.
Exited with code 100
```

mrshenli reviewed (commented):
This LGTM in general! Thanks! @pietern could you please help take another look?

The master is green now, let's try rebase again and see if that fixes the build error.

baobablyh commented:
> The master is green now, let's try rebase again and see if that fixes the build error.

Just rebased and let's see if how well CI can work this time. Thanks for taking a look!

Commit:
Add comment that local_used_maps_ need to be on the same device as replica

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -158,6 +158,20 @@ Reducer::Reducer(
+    local_used_maps_.resize(replica_count);
+    for (size_t i = 0; i < local_used_maps_.size(); i++) {
+      at::TensorOptions options;
+      options = options.device(replicas_[i][0].device());
```
Yes, that's the point. Just added comment.

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -529,6 +547,10 @@ void Reducer::prepare_for_backward(
   }
 
+  for (auto& local_used : local_used_maps_) {
+    local_used.fill_(1);
```
We need one separate 1D tensor for each replica, as each replica may be with different device.

Commit:
Use at::empty directly to create local_used_maps_ tensors

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -158,6 +158,20 @@ Reducer::Reducer(
+      options = options.device(replicas_[i][0].device());
+      options = options.dtype(at::kInt);
+      local_used_maps_[i] =
+          torch::autograd::make_variable(at::empty({variable_count}, options));
```
Good catch! at::empty is sufficient. Fixed.

baobablyh commented:
pytorch_linux_xenial_py2_7_9_test failure should be irrelevant, as it failed on master for same issue too:

https://circleci.com/gh/pytorch/pytorch/3589357?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link/console

Commit:
Fix type cast error

torch/csrc/distributed/c10d/reducer.cpp:174:40: error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'long' in initializer list [-Wc++11-narrowing]
      local_used_maps_[i] = at::empty({variable_count}, options);
                                       ^~~~~~~~~~~~~~
stderr: torch/csrc/distributed/c10d/reducer.cpp:174:40: error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'long' in initializer list [-Wc++11-narrowing]
      local_used_maps_[i] = at::empty({variable_count}, options);

baobablyh commented:
> Internal tests failed with the following error:
> 
> ```
> Summary: 
> torch/csrc/distributed/c10d/reducer.cpp:174:40: error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'long' in initializer list [-Wc++11-narrowing]
>       local_used_maps_[i] = at::empty({variable_count}, options);
>                                        ^~~~~~~~~~~~~~
> stderr: torch/csrc/distributed/c10d/reducer.cpp:174:40: error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'long' in initializer list [-Wc++11-narrowing]
>       local_used_maps_[i] = at::empty({variable_count}, options);
> ```

Fixed. Going forward, how can I capture this type of issue on my side? I did run clang-format however it didn't report anything.

baobablyh commented:
> > I can do it after this PR. Please let me know the thought. I would be OK with either one. Just don't want to block this PR for too long. Thanks!
> 
> +1 for doing it in a separate PR on top of this ([ghstack](https://github.com/ezyang/ghstack)), as long as we are committed to this and make sure we don't break existing use cases.
> 
> @pietern does the plan of having a separate function sounds OK to you?

Since there is no objection, I'm going to post another PR for 'no_sync' case with the same ticket #28294. Please let me know any concern. @pietern @mrshenli

baobablyh commented:
> > > I can do it after this PR. Please let me know the thought. I would be OK with either one. Just don't want to block this PR for too long. Thanks!
> > 
> > 
> > +1 for doing it in a separate PR on top of this ([ghstack](https://github.com/ezyang/ghstack)), as long as we are committed to this and make sure we don't break existing use cases.
> > @pietern does the plan of having a separate function sounds OK to you?
> 
> Since there is no objection, I'm going to post another PR for 'no_sync' case with the same ticket #28294. Please let me know any concern. @pietern @mrshenli

I thought about this again. I think we can handle 'no_sync' case with this PR. Going to post commit soon.

baobablyh commented:
> > > > I can do it after this PR. Please let me know the thought. I would be OK with either one. Just don't want to block this PR for too long. Thanks!
> > > 
> > > 
> > > +1 for doing it in a separate PR on top of this ([ghstack](https://github.com/ezyang/ghstack)), as long as we are committed to this and make sure we don't break existing use cases.
> > > @pietern does the plan of having a separate function sounds OK to you?
> > 
> > 
> > Since there is no objection, I'm going to post another PR for 'no_sync' case with the same ticket #28294. Please let me know any concern. @pietern @mrshenli
> 
> I thought about this again. I think we can handle 'no_sync' case with this PR. Going to post commit soon.

Tried to make it work however this case turns out much more complicated than I thought. Looks like we have to maintain states for unused params on both no_sync session level and iteration level. The currently logic with autograd_hook only supports iteration level. We have to refactor the existing code very carefully to add support for no_sync session level. So my suggestion is still handling no_sync case separately, unless we have some quick solution. @pietern

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -249,8 +266,12 @@ void Reducer::autograd_hook(VariableIndex index) {
     has_marked_unused_parameters_ = true;
+
+    // Allreduce locally used param maps to get the global consensus
+    local_used_work_ = process_group_->allreduce(local_used_maps_);
```
IIUC, @pietern has a concern that this might not work for `no_sync` mode, because `no_sync` mode does not call `prepare_for_backward` for intermediate outputs. When finally calling `prepare_for_backward` on the last output, it might have a different set of unused parameters compared to previous intermediate outputs. So the allreduce here might not compute the correct set of globally unused params.

However, it looks like the same problem already exist in today's DDP, where `prepare_for_backward` is only called on the last output. If `find_unused_parameters` is true, it won't correctly detect unused params either. It seems to me adding global unused support here won't break existing use cases? @pietern

mrshenli commented:
@pytorchbot rebase this please

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -238,6 +255,16 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
+  //  mark_unused_params(??);
+  //  // Where to reset this flag to false? post-backward hook?
+  //  called_flag = true;
+  //}
```
Chatted with @pietern offline yesterday, below is my understanding of the concern and the recommended approach:


#### Concern

The main concern is that we might have a regression on the `no_sync` mode if we don't implement it in this PR. Assume the application runs two iterations with `no_sync` mode:

1. the first iteration touches all parameters
2. the second iteration touches the first 10 parameters in the model

In this case, after exiting the `no_sync` mode, `prepare_for_backward` is called, and DDP hooks would start to trigger allreduce operations. However, only 10 of the autograd hook would fire in the second backward, and DDP would think all remaining parameters are unused and thus set them to None. This would result in incorrect result.

#### Proposal

To address this issue, one approach is to always update the bitmap in all backward passes, regardless of it's `no_sync` or not. It might be something like:

1. In the autograd hook, set `local_used_maps_ [index] = true`;
2. In `finalize_backward` kick off the allreduce on `local_used_maps_` and compute globally unused params
3. In `finalize_backward` after consuming `local_used_maps_`, reset them all to false/0.

One perf optimization might be kicking off the allreduce in `prepare_for_backward`, but that would mean we have to consider unused parameters in two different places:

* autograd hook in `no_sync` mode
* `prepare_for_backward` by walking through the autograd graph.

@pietern please correct me if I am wrong.

baobablyh replied:
Make sense to me. Will rework the last commit. Thanks!

Just a minor correction: with the current implementation we don't set the globally unused params to None. We just keep them untouched.

baobablyh replied:
Done. Thanks for the suggestion!

Commit:
Fix no_sync case

For no_sync session, we need to mark param used as long as it is ever used in any iteration during the session. In that way, we can get the correct params which are not used locally in the end of the session and then reduce the used maps to figure out the globally unused params during the entire no_sync session.

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -615,6 +649,10 @@ void Reducer::finalize_backward() {
   AT_ASSERT(next_bucket_ == buckets_.size());
 
+  // Allreduce locally used param maps to get the global consensus
+  local_used_work_ = process_group_->allreduce(local_used_maps_);
```
Looking tat this again, I wonder whether we can kick of the allreduce op in `mark_variable_read` ([here](https://github.com/pytorch/pytorch/blob/1ab2f043babb54a241e74db794b21f032478dc72/torch/csrc/distributed/c10d/reducer.cpp#L330)) and only wait for the op in this function?

pietern replied:
+1, that would save some time.

Also, I think we should keep the local version of `local_used_maps_` around. We know that if the variable was used locally, it is also used globally. Therefore, we don't need to wait for the reduction of `local_used_maps_` to complete before assigning/copying grads, for variables that are used locally. If instead we lazily wait for this work to complete, only when we see a variable that was unused locally, then we end up delaying the synchronization point that `local_used_work_->wait()` implies. If we don't have any unused parameters at all, we can skip waiting for the work to complete altogether, and cause negligible performance overhead for models where all parameters are used.

What do you think?

mrshenli replied:
> If instead we lazily wait for this work to complete, only when we see a variable that was unused locally

This would be a great perf optimization

baobablyh replied:
Yes, very nice idea for optimization. Thanks! @pietern 
Will do.

baobablyh replied:
> Looking tat this again, I wonder whether we can kick of the allreduce op in `mark_variable_read` ([here](https://github.com/pytorch/pytorch/blob/1ab2f043babb54a241e74db794b21f032478dc72/torch/csrc/distributed/c10d/reducer.cpp#L330)) and only wait for the op in this function?

Nice point! Done as [c37bb20](https://github.com/pytorch/pytorch/pull/28883/commits/c37bb2017f59cc08b1f744cfd121300488c85cb1)

baobablyh replied:
> Yes, very nice idea for optimization. Thanks! @pietern
> Will do.

Done as 5893be8.

Hope you don't mind I reused what you mentioned here in comments. Awesome optimization! Thanks!

mrshenli reviewed (commented):
Thanks! The algorithm looks correct to me! Build failures are real though:

```
Nov 22 00:36:11              ^~~~~~~~~~~~~~~~~~~
Nov 22 00:36:11 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/c10d/reducer.cpp.o
Nov 22 00:36:26 /var/lib/jenkins/workspace/torch/csrc/distributed/c10d/reducer.cpp: In member function 'void c10d::Reducer::finalize_bucket_dense(c10d::Reducer::Bucket&)':
Nov 22 00:36:26 /var/lib/jenkins/workspace/torch/csrc/distributed/c10d/reducer.cpp:607:53: error: 'class at::Tensor' has no member named 'grad_accumulator'
Nov 22 00:36:26        const VariableIndex& var_idx = func_[variable.grad_accumulator().get()];
```

There are some changes in `reducer.cpp` four days ago. Could you please rebase and address the build failure? Thanks!

pietern commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -579,13 +603,23 @@ void Reducer::finalize_bucket_dense(Bucket& bucket) {
+      const VariableIndex& var_idx = func_[variable.grad_accumulator().get()];
+      bool global_unused =
+          local_used_maps_[var_idx.replica_index][var_idx.variable_index]
+              .item<int>() == 0;
```
This tensor should be copied to CPU once and used many times. If this sits on GPU then this type of access will cause many GPU-to-CPU copies and add considerable overhead.

pietern replied:
Regarding the replica_index/variable_index: could you consider adding a vector of `VariableIndex` to the bucket replace so we don't need to mess with the grad accumulator pointer here? Would be a bit cleaner.

baobablyh replied:
> Regarding the replica_index/variable_index: could you consider adding a vector of `VariableIndex` to the bucket replace so we don't need to mess with the grad accumulator pointer here? Would be a bit cleaner.

Done as [043ea04](https://github.com/pytorch/pytorch/pull/28883/commits/043ea04c1de49ebf39186950513234a7e5bfd0c4)

baobablyh replied:
> This tensor should be copied to CPU once and used many times. If this sits on GPU then this type of access will cause many GPU-to-CPU copies and add considerable overhead.

Good point. Fixed as [https://github.com/pytorch/pytorch/pull/28883/commits/b12597171fb70f3d2c183b479876740c30d78f4a](b125971)

pietern reviewed (commented):
This is great work, thank you @baobablyh!

Lazily waiting on the unused maps reduction means minimizing performance impact for the big majority of models where all parameters are always used. Then we only pay the overhead cost if there is indeed a parameter that is locally unused, because we need to check if it's also *globally* unused.

Commit:
Fix issue caused by recent API change

variable.grad_accumulator() -> torch::autograd::impl::grad_accumulator(variable)

baobablyh commented:
> Thanks! The algorithm looks correct to me! Build failures are real though:
> 
> ```
> Nov 22 00:36:11              ^~~~~~~~~~~~~~~~~~~
> Nov 22 00:36:11 [100%] Building CXX object caffe2/torch/CMakeFiles/torch_python.dir/csrc/distributed/c10d/reducer.cpp.o
> Nov 22 00:36:26 /var/lib/jenkins/workspace/torch/csrc/distributed/c10d/reducer.cpp: In member function 'void c10d::Reducer::finalize_bucket_dense(c10d::Reducer::Bucket&)':
> Nov 22 00:36:26 /var/lib/jenkins/workspace/torch/csrc/distributed/c10d/reducer.cpp:607:53: error: 'class at::Tensor' has no member named 'grad_accumulator'
> Nov 22 00:36:26        const VariableIndex& var_idx = func_[variable.grad_accumulator().get()];
> ```
> 
> There are some changes in `reducer.cpp` four days ago. Could you please rebase and address the build failure? Thanks!

Done. Thanks!

baobablyh commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -74,6 +74,14 @@ class Reducer {
+  // Locally used parameter maps indicating if parameters are used locally
+  // during the current iteration. One tensor for each model replica and each
+  // tensor is one-dim int32 tensor of number of parameters. These tensors are
+  // marked and allreduced at the end of forward for figuring out the globally
```
Yes. Will correct.

baobablyh replied:
Done.

Commit:
Address some comments

* Remove irrelevant new line
* Update the comment for local_used_maps_

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -251,6 +274,7 @@ void Reducer::autograd_hook(VariableIndex index) {
   // function and their indexes stored in the `unused_parameters_` vector.
   if (!has_marked_unused_parameters_ && !unused_parameters_.empty()) {
     has_marked_unused_parameters_ = true;
+
```
Done.

Commit:
Add per-bucket vector for global indice of participating variables in bucket

Commit:
Kick off reduction for local_used_maps_ once the final bucket gets marked ready

Commit:
Use CPU tensors for bookkeeping locally used params and GPU tensors for reduction

Commit:
Use in-place copy for D2H and H2D local used maps

Commit:
Lazily wait for local_unused_map reduction to complete

baobablyh commented:
@pietern @mrshenli 

Looks like I have addressed all the comments. Please let me know if it looks good or I missed anything. Thanks!

Commit:
Initialize local_used_maps_reduced_ to false

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -472,6 +507,8 @@ void Reducer::initialize_buckets(
           .intra_bucket_index = intra_bucket_index++,
       };
+
+      bucket.variable_indices.push_back(variable_index);
```
`bucket.variable_indice`'s content is the same as `bucket_indices[bucket_index]`? If so, is it OK to move the latter into the former?

baobablyh replied:
Good point. Done as cc28ee6.

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -572,20 +609,56 @@ void Reducer::prepare_for_backward(
+      // a parameter that is locally unused, because we need to check if it's
+      // also globally unused.
+      size_t variable_index = bucket.variable_indices[intra_bucket_index];
+      bool global_unused =
```
Let's add a comment here to mention that the var `global_unused` might not be global yet. It is only global if `local_used_maps_reduced_  == true`.

baobablyh replied:
Done with cc28ee6.

mrshenli commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -572,20 +609,56 @@ void Reducer::prepare_for_backward(
 
 // A bucket with one or more dense tensors needs to be unflattened.
 void Reducer::finalize_bucket_dense(Bucket& bucket) {
-  for (auto& replica : bucket.replicas) {
```
Shall we add an `TORCH_INTERNAL_ASSERT`/`AT_ERROR` that `local_used_maps_reduced_ ` must be false when entering this method.

baobablyh replied:
Done with cc28ee6.

mrshenli reviewed (approved):
This looks great to me! Thanks a lot for contributing! I added some minor comments below.

pietern commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -572,20 +609,56 @@ void Reducer::prepare_for_backward(
+      size_t variable_index = bucket.variable_indices[intra_bucket_index];
+      bool global_unused =
+          local_used_maps_[replica_index][variable_index].item<int>() == 0;
+      if (global_unused && !local_used_maps_reduced_) {
```
We could use the presence or absence of a work object in the `shared_ptr<Work>` as a proxy for the boolean `local_used_maps_reduced_`.

Then, before kicking off the *next* allreduce, we can call `local_used_work_->wait()` to avoid the contents of `local_used_maps_dev_` to be clobbered by a previous allreduce in the next iteration (which is possible if there are multiple in flight operations against the same tensor).

baobablyh replied:
I don't feel very comfortable with such implicit indication and would rather use a variable (local_used_maps_reduced_) to explicitly and clearly indicate if local_used_maps has been reduced or not, unless you insist very strongly.

pietern replied:
That's fine, we can keep it as is. The second part of the comment still holds, though.

baobablyh replied:
OK. I'll address the 2nd part.

baobablyh replied:
Done with bc1c493.

pietern commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -625,6 +698,12 @@ void Reducer::finalize_backward() {
+  for (auto& local_used : local_used_maps_) {
+    local_used.fill_(0);
+  }
+  local_used_maps_reduced_ = false;
```
The work object needs to be cleaned up as well, but like I said in the previous comment, it's probably best to do this lazily if it wasn't needed in the first place. Then the only overhead we end up paying is the H2D memcpy and the kernel launch cost.

baobablyh replied:
Not sure if I'm following ... We already do lazy wait as you suggested previously. For example, if all params are used locally, we just skip work wait and D2H copy. For such case, the overhead is just H2D copy and the allreduce kernel launch cost.

pietern replied:
The difference with the bucket work is that we wait on all of them.

If we don't wait, it is possible for the next iteration to interfere and write to the device-side memory for `local_unused_maps_dev_` while the allreduce of the previous iteration is still going. Sure, it's a long shot, but we must eliminate the possibility of this happening at all.

baobablyh replied:
I see. That's valid point. I'll address. Thanks @pietern .

baobablyh replied:
Done with bc1c493.

pietern reviewed (commented):
Thanks, @baobablyh! This is turning into an excellent change.

Because this change touches the hot path, I'd like to double check the performance implication before we merge. Let's check some basic image models and some big NLP models to cover our bases. I anticipate the cost to be negligible but want to be sure.

baobablyh commented:
> Thanks, @baobablyh! This is turning into an excellent change.
> 
> Because this change touches the hot path, I'd like to double check the performance implication before we merge. Let's check some basic image models and some big NLP models to cover our bases. I anticipate the cost to be negligible but want to be sure.

Is there any existing test case covering such perf implication? how do we run such things usually?

Commit:
Address review comments

baobablyh commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -144,6 +144,9 @@ class Reducer {
   struct Bucket {
     std::vector<BucketReplica> replicas;
 
+    // Global indice of participating variables in the bucket
```
Fixed with cc28ee6.

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -158,6 +160,27 @@ Reducer::Reducer(
+    local_used_maps_dev_.resize(replica_count);
+
+    for (size_t i = 0; i < replica_count; i++) {
+      local_used_maps_[i] = at::zeros({static_cast<long>(variable_count)});
```
Nice catch. Fixed with cc28ee6.

baobablyh commented:
> > Thanks, @baobablyh! This is turning into an excellent change.
> > Because this change touches the hot path, I'd like to double check the performance implication before we merge. Let's check some basic image models and some big NLP models to cover our bases. I anticipate the cost to be negligible but want to be sure.
> 
> @pietern Is there any existing test case covering such perf implication? how do we run such things usually?

I ran our internal models which are large and did not observe anything obvious. If you would like me to validate other models, could you please send me more detailed information? Thanks. @pietern

pietern commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -609,6 +608,10 @@ void Reducer::prepare_for_backward(
+  AT_ASSERTM(
+      !local_used_maps_reduced_,
+      "local_used_maps should have not been reduced yet at this point");
+
```
This may trigger. This function is called once per bucket, so if the first one waits for the reduction and sets `local_used_maps_reduced_`, then the second bucket will trip this assert.

Also please use `TORCH_INTERNAL_ASSERT` instead of `AT_ASSERTM`. The latter is being deprecated.

baobablyh replied:
Ah, that's right. I'll remove the assert then.

baobablyh replied:
Done with bc1c493

meiluzhu commented:
Hi, I know the synchronization of gradients happens during the backwards for DDP. But I didn’t find the corresponding script in backwards. Where can I find it?

baobablyh commented:
> Hi, I know the synchronization of gradients happens during the backwards for DDP. But I didn’t find the corresponding script in backwards. Where can I find it?

Please ask such question on "questions" slack channel. We usually don't use PR for Q/A. Thanks for understanding.

Commit:
Address comments

baobablyh commented:
> > Thanks, @baobablyh! This is turning into an excellent change.
> > Because this change touches the hot path, I'd like to double check the performance implication before we merge. Let's check some basic image models and some big NLP models to cover our bases. I anticipate the cost to be negligible but want to be sure.
> 
> @pietern Is there any existing test case covering such perf implication? how do we run such things usually?

Hi Pieter, could you share more information about the perf validation you mentioned? such as what particular models? what test configs? how to run them? etc. I ran some model and perf looks good to me. If you would like me to run more perf validation, could you please provide further information? Thanks.

baobablyh commented:
> > > Thanks, @baobablyh! This is turning into an excellent change.
> > > Because this change touches the hot path, I'd like to double check the performance implication before we merge. Let's check some basic image models and some big NLP models to cover our bases. I anticipate the cost to be negligible but want to be sure.
> > 
> > 
> > @pietern Is there any existing test case covering such perf implication? how do we run such things usually?
> 
> Hi Pieter, could you share more information about the perf validation you mentioned? such as what particular models? what test configs? how to run them? etc. I ran some model and perf looks good to me. If you would like me to run more perf validation, could you please provide further information? Thanks.

@pietern @mrshenli 

I ran the further perf validation with 16 GPUs on 2 nodes and noticed about 10% perf degradation. The issue is the blocking H2D copy taking too long:
`local_used_maps_dev_[i].copy_(local_used_maps_[i])`

Thanks a lot for the suggestion of validating perf before merging! which does capture the issue.

I tried to use pinned memory and async copy which solved the perf degradation:
```
at::zeros({static_cast<long>(variable_count)}, options.pinned_memory(true));
...
local_used_maps_dev_[i].copy_(local_used_maps_[i], true);
local_used_work_ = process_group_->allreduce(local_used_maps_dev_);
```
However I'm not sure if we can assume that async H2D and the subsequent allreduce use the same stream? If so, we should be fine as allreduce kernel execution will start after async H2D completes as they are in the same stream. If not, we may be in trouble ...

Thought? Thanks!

pietern commented:
Thanks for verifying and coming up with the fix.

The async copy and the allreduce respect the current stream, so they will be sequenced correctly.

Can you confirm if there is still any perf impact remaining with the fix enabled?

Commit:
Use async H2D copy to avoid the blocking overhead

baobablyh commented:
> Thanks for verifying and coming up with the fix.
> 
> The async copy and the allreduce respect the current stream, so they will be sequenced correctly.
> 
> Can you confirm if there is still any perf impact remaining with the fix enabled?

Thanks for confirming. With resnet50, the H2D part has been fixed by afc9aab from what I can tell. I ran a bunch of experiments and did not observe the difference between the branch and master.

However, D2H part is a blocking call as well so has the similar impact. Looks like we have done our best to do that unless really necessary (thanks to the lazy wait!). Any thought for optimizing out D2H? One way I can think of is to kick off async D2H copy right after allreduce kernel gets launched, which may require some change in processgroup and backend.

baobablyh commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -172,7 +172,7 @@ Reducer::Reducer(
       options = options.dtype(at::kInt);
       local_used_maps_[i] =
-          at::zeros({static_cast<long>(variable_count)}, options);
+          at::zeros({static_cast<long>(variable_count)}, options.pinned_memory(true));
```
Hitting this in CI tests. How should we handle this properly? @pietern @mrshenli Thanks.
```
Nov 26 06:19:00   File "test_c10d.py", line 2937, in test_single_dtype_single_bucket
Nov 26 06:19:00     dist.Reducer([parameters], buckets, self.process_group)
Nov 26 06:19:00 RuntimeError: Pinned memory requires CUDA. PyTorch splits its backend into two shared libraries: a CPU library and a CUDA library; this error has occurred because you are trying to use some CUDA functionality, but the CUDA library has not been loaded by the dynamic linker for some reason.  The CUDA library MUST be loaded, EVEN IF you don't directly use any symbols from the CUDA library! One common culprit is a lack of -Wl,--no-as-needed in your link arguments; many dynamic linkers will delete dynamic library dependencies if you don't depend on any of their symbols.  You can check if this has occurred by using ldd on your binary to see if there is a dependency on *_cuda.so library. (getPinnedMemoryAllocator at ../aten/src/ATen/detail/CUDAHooksInterface.h:107)
```

pietern replied:
I think it will be fine if you make `options.pinned_memory(true)` conditional on `replicas_[i][0].is_cuda()`. The asynchronous copy becomes synchronous if you're copying from CPU tensor to CPU tensor, so you can keep the rest as is.

baobablyh replied:
Thanks. Done with 5ec6672. Originally I thought there is some global config for this purpose.

Commit:
Allocate local_used_maps as pinned memory if cuda is enabled

baobablyh commented:
Rebased.

baobablyh commented:
> Rebased.

CI test failures are caused by this one: #30465

baobablyh commented:
Rebased.

baobablyh commented:
@pietern @mrshenli 

03262f6 is a draft commit for optimizing out the blocking D2H copy. It works well and in my test fixed the perf degradation caused by D2H. Could you take a quick look? Just want to make sure I'm on the right direction before I refine it further. Thanks! I'm now satisfied with the optimization we put here ...

baobablyh commented:
> @pietern @mrshenli
> 
> [03262f6](https://github.com/pytorch/pytorch/commit/03262f6fb84b6b9eaf37a8f90b56e3470a8d7bfb) is a draft commit for optimizing out the blocking D2H copy. It works well and in my test fixed the perf degradation caused by D2H. Could you take a quick look? Just want to make sure I'm on the right direction before I refine it further. Thanks! I'm now satisfied with the optimization we put here ...

Looks like we can get rid of local_used_maps_dev_ and end up with using two pinned memory buffers. Running more tests and hope it will work.

baobablyh commented:
@pietern @mrshenli 

With d649744, I use pinned memory for allreduce. Since dev can access pinned memory directly, we don't need explicit D2H/H2D anymore.

I validated the perf with resnet50. The validation ran well. It's comparable with master.

pietern commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -572,20 +606,62 @@ void Reducer::prepare_for_backward(
+        local_used_work_->wait();
+        // At this point, local_used_maps_2_ has the global consensus. We swap
+        // their content.
+        std::swap(local_used_maps_1_, local_used_maps_2_);
```
You can't do this yet here. NCCL collectives run asynchronously and the `wait()` call only synchronizes them with the current CUDA stream. This means only CUDA kernels will be sequenced after the NCCL kernel. Because you're accessing host memory here, it may still be running, and therefore get an incorrect result. You need to perform device synchronization to be certain you can access it.

baobablyh replied:
Yes, you are right. Host memory may not be ready yet at this point.

pietern commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -625,6 +701,22 @@ void Reducer::finalize_backward() {
+  for (size_t i = 0; i < local_used_maps_1_->size(); i++) {
+    (*local_used_maps_1_)[i].fill_(0);
+    (*local_used_maps_2_)[i].fill_(0);
+  }
```
This needs to happen after you have waited on the work (in case lazy waiting never happened), because otherwise NCCL may still be mutating this memory until after you have filled it with 0's.

pietern reviewed (commented):
Thanks for continuing to optimize this, @baobablyh.

Per the reasons I mentioned in the comments, I think it's best to revert to the previous state and keep a device-side copy of the `local_used_maps_`. Then can reason about sequencing and completion more easily. Sure, it means triggering a blocking DtoH copy once we need the result, but I don't think we can avoid it anyway.

We'll end up with:
* Setting the used map in CPU memory to 1 when a gradient was computed
* After all gradients were computed, we trigger a H2D copy, and an allreduce
* During finalization, we use the local used map in CPU memory to figure out if we need to wait
* If we do, wait on the local used work, trigger a blocking D2H copy (this is expensive), swap vectors
* If we never do, wait on the local used work, but don't trigger a D2H copy.
* We fill the host side local used map with 0's

The blocking D2H copy is what makes this thing expensive and can cause the overlap between grad computation and reduction to be gone (because the allreduce is sequenced after all other allreduces). But I don't think there is another way to deal with this, without running excessively more reductions.

At least, in this proposed solution, we'll keep the overhead to an absolute minimum for the cases where there are no unused parameters.

baobablyh commented:
> Per the reasons I mentioned in the comments, I think it's best to revert to the previous state and keep a device-side copy of the `local_used_maps_`. Then can reason about sequencing and completion more easily. Sure, it means triggering a blocking DtoH copy once we need the result, but I don't think we can avoid it anyway.

@pietern 

Thanks for the suggestion which makes a lot of sense. I just reverted the last commit for optimizing out D2H. Yes, it seems hard to get rid of it. I'll stop optimization right here. Let's keep D2H for now as you mentioned.

As we discussed, this PR is in good mergable state. The only pending thing is the perf eval you are going to do before merging. Could you please keep us posted on the eval result once we have it? Thanks!

baobablyh commented:
Rebased.

baobablyh commented:
> @mrshenli has imported this pull request. If you are a Facebook employee, you can view this diff [on Phabricator](https://our.internmc.facebook.com/intern/diff/D18491530/).

Thanks for the help! @mrshenli

pietern commented:
I wrote a benchmark tool to sweep over whatever number of processes that is allocated and run a throughput test on a number of models (that don't use unused parameters), to evaluate if there is any runtime impact for use cases where this doesn't apply. I'll submit this as a PR soon, but wanted to share the results of this test first. I ran it on 8 DGX machines with 8x V100 GPUs each.

The results show that this PR doesn't have any performance impact for models that don't have unused parameters. For models that do, I expect there to be a performance impact, but we would need to add them to the tool to quantify it.

```
                                 baseline                      test
                     --------------------      --------------------
cuda_version:                        10.0  vs                  10.0
pytorch_version:          1.4.0a0+fcb7371  vs       1.4.0a0+a84cca8

Benchmark: resnet50 with batch size 32
   1 GPUs: p75:  0.097 (-0.0%)  p95:  0.097 (-0.0%)
   2 GPUs: p75:  0.103 (-0.0%)  p95:  0.103 (-0.0%)
   4 GPUs: p75:  0.103 (-0.0%)  p95:  0.103 (-0.0%)
   8 GPUs: p75:  0.104 (-0.0%)  p95:  0.106 (+0.0%)
  16 GPUs: p75:  0.105 (-0.0%)  p95:  0.107 (+0.0%)
  24 GPUs: p75:  0.105 (-0.0%)  p95:  0.106 (+0.0%)
  32 GPUs: p75:  0.106 (+0.0%)  p95:  0.106 (+0.0%)
  40 GPUs: p75:  0.107 (+0.0%)  p95:  0.108 (+0.0%)
  48 GPUs: p75:  0.106 (+0.0%)  p95:  0.108 (+0.0%)
  56 GPUs: p75:  0.107 (+0.0%)  p95:  0.109 (+0.0%)
  64 GPUs: p75:  0.106 (-0.0%)  p95:  0.106 (-0.0%)

Benchmark: resnet101 with batch size 32
   1 GPUs: p75:  0.162 (-0.0%)  p95:  0.162 (-0.0%)
   2 GPUs: p75:  0.176 (-0.0%)  p95:  0.177 (-0.0%)
   4 GPUs: p75:  0.178 (+0.0%)  p95:  0.179 (+0.0%)
   8 GPUs: p75:  0.180 (+0.0%)  p95:  0.180 (+0.0%)
  16 GPUs: p75:  0.181 (+0.0%)  p95:  0.182 (-0.0%)
  24 GPUs: p75:  0.181 (-0.0%)  p95:  0.190 (+0.0%)
  32 GPUs: p75:  0.186 (+0.0%)  p95:  0.192 (-0.0%)
  40 GPUs: p75:  0.188 (+0.0%)  p95:  0.192 (+0.0%)
  48 GPUs: p75:  0.182 (-0.0%)  p95:  0.188 (-0.0%)
  56 GPUs: p75:  0.182 (+0.0%)  p95:  0.183 (-0.0%)
  64 GPUs: p75:  0.182 (-0.0%)  p95:  0.186 (-0.0%)

Benchmark: resnext50_32x4d with batch size 32
   1 GPUs: p75:  0.145 (-0.0%)  p95:  0.145 (-0.0%)
   2 GPUs: p75:  0.153 (-0.0%)  p95:  0.153 (-0.0%)
   4 GPUs: p75:  0.153 (-0.0%)  p95:  0.153 (-0.0%)
   8 GPUs: p75:  0.157 (-0.0%)  p95:  0.157 (-0.0%)
  16 GPUs: p75:  0.159 (-0.0%)  p95:  0.159 (-0.0%)
  24 GPUs: p75:  0.159 (-0.0%)  p95:  0.159 (-0.0%)
  32 GPUs: p75:  0.160 (-0.0%)  p95:  0.161 (+0.0%)
  40 GPUs: p75:  0.160 (-0.0%)  p95:  0.160 (-0.0%)
  48 GPUs: p75:  0.160 (+0.0%)  p95:  0.160 (+0.0%)
  56 GPUs: p75:  0.161 (-0.0%)  p95:  0.161 (+0.0%)
  64 GPUs: p75:  0.160 (-0.0%)  p95:  0.162 (-0.0%)

Benchmark: resnext101_32x8d with batch size 32
   1 GPUs: p75:  0.416 (+0.0%)  p95:  0.417 (-0.0%)
   2 GPUs: p75:  0.439 (-0.0%)  p95:  0.440 (-0.0%)
   4 GPUs: p75:  0.443 (+0.0%)  p95:  0.444 (-0.0%)
   8 GPUs: p75:  0.450 (-0.0%)  p95:  0.451 (-0.0%)
  16 GPUs: p75:  0.455 (-0.0%)  p95:  0.456 (-0.0%)
  24 GPUs: p75:  0.456 (+0.0%)  p95:  0.457 (-0.0%)
  32 GPUs: p75:  0.457 (-0.0%)  p95:  0.458 (-0.0%)
  40 GPUs: p75:  0.460 (-0.0%)  p95:  0.460 (-0.0%)
  48 GPUs: p75:  0.460 (-0.0%)  p95:  0.461 (-0.0%)
  56 GPUs: p75:  0.463 (+0.0%)  p95:  0.464 (-0.0%)
  64 GPUs: p75:  0.460 (-0.0%)  p95:  0.462 (-0.0%)
```

baobablyh commented:
Rebased.

baobablyh commented:
> I wrote a benchmark tool to sweep over whatever number of processes that is allocated and run a throughput test on a number of models (that don't use unused parameters), to evaluate if there is any runtime impact for use cases where this doesn't apply. I'll submit this as a PR soon, but wanted to share the results of this test first. I ran it on 8 DGX machines with 8x V100 GPUs each.
> 
> The results show that this PR doesn't have any performance impact for models that don't have unused parameters. For models that do, I expect there to be a performance impact, but we would need to add them to the tool to quantify it.
> 

@pietern Very nice benchmarking tool! It will be very helpful going forward and we should run it whenever some change like this PR comes up to gate the perf regression. :+1:

baobablyh commented:
rebased

mrshenli commented on test/test_c10d.py:
```diff
@@ -2518,6 +2518,76 @@ def test_find_unused_parameters(find_unused_parameters, test_default=False):
+        class GlobalLocalUnusedParamModule(nn.Module):
+            class Task(nn.Module):
+                def __init__(self):
+                    super().__init__()
```
we haven't deprecated py2 support yet. So use `super(Task, self).__init__()` here?

baobablyh replied:
Good suggestion. Will make the change.

baobablyh replied:
Done with 1bc27e3.

Commit:
Use python2 style for super call

kostmo commented:
## CircleCI build failures summary
As of commit 8fe4782:
* **1/1** broken upstream at merge base d4c25ad (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=d4c25add458f65fb9de1f1310cd536d4c073d3aa&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * Since your merge base is older than `viable/strict`, run these commands:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8fe4782c370e3214fae78f2caf8d34302ad94663).

### 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_linux_xenial_py3_6_gcc5_4_build](https://dr.pytorch.org/build-details.html?build_id=73668759)



---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

baobablyh commented:
Rebased.

Commit:
Use device guard to enclose pin memory allocation
